In [14]:
import pandas as pd
import numpy as np
import urllib.request
from bs4 import BeautifulSoup

In [15]:
z='''Africa
Botswana  
Cameroon
Ethiopia
Eritrea
The Gambia
Ghana
Kenya
Lesotho
Liberia
Malawi
Mauritius
Namibia
Nigeria
Rwanda
Seychelles
Sierra Leone
South Africa
South Sudan
Sudan
Swaziland
Tanzania  
Uganda
Zambia
Zimbabwe'''

In [16]:
africa_eng=z.split("\n")

In [17]:
y='''Antigua and Barbuda
The Bahamas
Barbados
Belize  
Canada
Dominica
Grenada
Guyana
Jamaica
Saint Kitts and Nevis
Saint Lucia
Saint Vincent and the Grenadines
Trinidad and Tobago
United States '''

In [18]:
americas_eng=y.split('\n')

In [19]:
americas_eng

['Antigua and Barbuda',
 'The Bahamas',
 'Barbados',
 'Belize  ',
 'Canada',
 'Dominica',
 'Grenada',
 'Guyana',
 'Jamaica',
 'Saint Kitts and Nevis',
 'Saint Lucia',
 'Saint Vincent and the Grenadines',
 'Trinidad and Tobago',
 'United States ']

In [20]:
x='''India
Pakistan
Philippines
Singapore'''

In [21]:
asia=x.split('\n')

In [22]:
asia

['India', 'Pakistan', 'Philippines', 'Singapore']

In [23]:
w='''Australia
Fiji
Kiribati
Marshall Islands
Federated States of Micronesia
Nauru
New Zealand
Palau  
Papua New Guinea
Samoa
Solomon Islands
Tonga
Tuvalu
Vanuatu'''

In [24]:
oceania_eng=w.split('\n')

In [25]:
oceania_eng

['Australia',
 'Fiji',
 'Kiribati',
 'Marshall Islands',
 'Federated States of Micronesia',
 'Nauru',
 'New Zealand',
 'Palau  ',
 'Papua New Guinea',
 'Samoa',
 'Solomon Islands',
 'Tonga',
 'Tuvalu',
 'Vanuatu']

In [26]:
v='''Ireland
Malta
United Kingdom  '''

In [27]:
europ_eng=v.split('\n')

In [28]:
europ_eng

['Ireland', 'Malta', 'United Kingdom  ']

# Joining English speeking contries

In [29]:
mod_eng=americas_eng+asia+oceania_eng+europ_eng # add the all lists

In [30]:
print(len(mod_eng),'\n', mod_eng)

35 
 ['Antigua and Barbuda', 'The Bahamas', 'Barbados', 'Belize  ', 'Canada', 'Dominica', 'Grenada', 'Guyana', 'Jamaica', 'Saint Kitts and Nevis', 'Saint Lucia', 'Saint Vincent and the Grenadines', 'Trinidad and Tobago', 'United States ', 'India', 'Pakistan', 'Philippines', 'Singapore', 'Australia', 'Fiji', 'Kiribati', 'Marshall Islands', 'Federated States of Micronesia', 'Nauru', 'New Zealand', 'Palau  ', 'Papua New Guinea', 'Samoa', 'Solomon Islands', 'Tonga', 'Tuvalu', 'Vanuatu', 'Ireland', 'Malta', 'United Kingdom  ']


# 3 dig code of english spk contries extracting from web

In [37]:
html='https://www.worldatlas.com/aatlas/ctycodes.htm'

In [38]:
html_doc=urllib.request.urlopen(html)

In [39]:
soup=BeautifulSoup(html_doc, 'html.parser')


In [40]:
table=soup.find('table')

In [41]:
#Generate lists
A=[]
B=[]
C=[]
D=[]
E=[]

for row in table.find_all("tr"):
    cells = row.find_all('td')
  
    if len(cells)==5: #Only extract table body not heading
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))
        D.append(cells[3].find(text=True))
        E.append(cells[4].find(text=True))
        

In [42]:
#import pandas to convert list to data frame
import pandas as pd
A.pop(0)
B.pop(0)
C.pop(0)
D.pop(0)
E.pop(0)
df=pd.DataFrame(A,columns=['Country'])
df['A2(ISO)']=B
df['A3']=C
df['NUM(UN)']=D
df['Dial(cod)']=E

df.head()

Country A2(ISO)   A3 NUM(UN) Dial(cod)
0     Afghanistan      AF  AFG       4        93
1         Albania      AL  ALB       8       355
2         Algeria      DZ  DZA      12       213
3  American Samoa      AS  ASM      16     1-684
4         Andorra      AD  AND      20       376

In [43]:
common=[]
for j in range(df.Country.count()):
    for i in mod_eng:
        if df.Country[j] ==i:
            common.append(df.A3[j])
print(len(common) ,'\n', common)

29 
 ['ATG', 'AUS', 'BRB', 'CAN', 'DMA', 'FJI', 'GRD', 'GUY', 'IND', 'IRL', 'JAM', 'KIR', 'MLT', 'MHL', 'NRU', 'NZL', 'PAK', 'PNG', 'PHL', 'KNA', 'LCA', 'VCT', 'WSM', 'SGP', 'SLB', 'TON', 'TTO', 'TUV', 'VUT']


In [54]:
'''from pandas import DataFrame

Numbers = {'set_of_numbers': [1,2,3,4,5,6,7,8,9,10]}
df = DataFrame(Numbers,columns=['set_of_numbers'])

df.loc[df.set_of_numbers <= 4, 'equal_or_lower_than_4?'] = 'True' 
df.loc[df.set_of_numbers > 4, 'equal_or_lower_than_4?'] = 'False' 

print (df)

::
df['equal_or_lower_than_4?'] = df['set_of_numbers'].apply(lambda x: 'True' if x <= 4 else 'False')'''

"from pandas import DataFrame\n\nNumbers = {'set_of_numbers': [1,2,3,4,5,6,7,8,9,10]}\ndf = DataFrame(Numbers,columns=['set_of_numbers'])\n\ndf.loc[df.set_of_numbers <= 4, 'equal_or_lower_than_4?'] = 'True' \ndf.loc[df.set_of_numbers > 4, 'equal_or_lower_than_4?'] = 'False' \n\nprint (df)\n\n::\ndf['equal_or_lower_than_4?'] = df['set_of_numbers'].apply(lambda x: 'True' if x <= 4 else 'False')"

# code of comparing DF with list elements

# comparing two dataframes

In [46]:
companies_table = pd.read_csv('./companies.csv')
companies_table.head()

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list funding_total_usd  \
0                                              Media          10000000   
1  Application Platforms|Real Time|Social Network...            700000   
2                                  Apps|Games|Mobile           3406878   
3                                        Curated Web           2000000   
4                                           Software                 -   

      status country_code state_code                 region           city  \
0  operating          IND         16                 Mumbai         Mumbai   
1  operating          USA         DE             DE - Other  Delaware City   
2  operating          NaN        NaN                    NaN            NaN   
3  operating          CHN         22                Beijing        Beijing   
4  operating          USA         IL  Springfield, Illinois      Champaign   

   funding_rounds  founded_at first_funding_at last_funding_at  
0               1         NaN       05-01-2015      05-01-2015  
1               2  04-09-2014       01-03-2014      14-10-2014  
2               1         NaN       30-01-2014      30-01-2014  
3               1  01-01-2007       19-03-2008      19-03-2008  
4               1  01-01-2010       24-07-2014      24-07-2014

In [47]:
companies_table['eng_spk'] =companies_table['country_code'].apply(lambda x:'True' if x in common else 'False')

companies_table['eng_spk'].value_counts()

False    61231
True      5137
Name: eng_spk, dtype: int64

## There are total 5137 companys in countries where english is an official language

In [48]:
eng_true=companies_table.where(companies_table['eng_spk']=='True')


- Top English speaking country
- Second English speaking country
- Third English speaking country


In [49]:

grouped = eng_true.groupby('country_code')

print(grouped.agg(np.size))

              permalink  name  homepage_url  category_list  funding_total_usd  \
country_code                                                                    
AUS                 503   503           503            503                503   
BRB                   1     1             1              1                  1   
CAN                1925  1925          1925           1925               1925   
DMA                   1     1             1              1                  1   
GRD                   1     1             1              1                  1   
IND                1596  1596          1596           1596               1596   
IRL                 440   440           440            440                440   
JAM                   3     3             3              3                  3   
KNA                   1     1             1              1                  1   
MLT                   9     9             9              9                  9   
NZL                  99    9

# Top English speaking country =CAN
## Second English speaking country =IND
### Third English speaking country = AUS

In [61]:
round_table=pd.read_csv('./rounds2.csv',encoding = "ISO-8859-1")
round_table.count()

company_permalink          114949
funding_round_permalink    114949
funding_round_type         114949
funding_round_code          31140
funded_at                  114949
raised_amount_usd           94959
dtype: int64

# to rename coloumn 

In [62]:
round_table1=round_table.rename(columns=({'company_permalink':'permalink'}))
permalink=[x.lower() for x in round_table1.permalink]
round_table1['permalink']=permalink
round_table1.head()

permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

# we concatinated the two csv files by there coloumn names
## JOINs can be performed with join() or merge(). By default, join() will join the DataFrames on their indices.Each method has parameters allowing you to specify the type of join to perform (LEFT, RIGHT, INNER, FULL) or the columns to join on (column names or indices).

In [63]:
mod_table=pd.merge(companies_table,round_table1, on='permalink',how ='outer')
mod_table.head(2)

permalink      name            homepage_url  \
0     /organization/-fame     #fame      http://livfame.com   
1  /organization/-qounter  :Qounter  http://www.qounter.com   

                                       category_list funding_total_usd  \
0                                              Media          10000000   
1  Application Platforms|Real Time|Social Network...            700000   

      status country_code state_code      region           city  \
0  operating          IND         16      Mumbai         Mumbai   
1  operating          USA         DE  DE - Other  Delaware City   

   funding_rounds  founded_at first_funding_at last_funding_at eng_spk  \
0             1.0         NaN       05-01-2015      05-01-2015    True   
1             2.0  04-09-2014       01-03-2014      14-10-2014   False   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN

In [64]:
mod_table.permalink.count()

115017

In [65]:
df.head()

Country A2(ISO)   A3 NUM(UN) Dial(cod) eng_spk
0     Afghanistan      AF  AFG       4        93   False
1         Albania      AL  ALB       8       355   False
2         Algeria      DZ  DZA      12       213   False
3  American Samoa      AS  ASM      16     1-684   False
4         Andorra      AD  AND      20       376   False

In [66]:
a = df['A3']

In [67]:
a

0      AFG
1      ALB
2      DZA
3      ASM
4      AND
5      AGO
6      AIA
7      ATA
8      ATG
9      ARG
10     ARM
11     ABW
12     AUS
13     AUT
14     AZE
15     BHS
16     BHR
17     BGD
18     BRB
19     BLR
20     BEL
21     BLZ
22     BEN
23     BMU
24     BTN
25     BOL
26     BES
27     BIH
28     BWA
29     BVT
      ... 
218    TZA
219    THA
220    TLS
221    TGO
222    TKL
223    TON
224    TTO
225    TUN
226    TUR
227    TKM
228    TCA
229    TUV
230    UGA
231    UKR
232    ARE
233    GBR
234    USA
235    UMI
236    URY
237    UZB
238    VUT
239    VEN
240    VNM
241    VGB
242    VIR
243    WLF
244    ESH
245    YEM
246    ZMB
247    ZWE
Name: A3, Length: 248, dtype: object

In [68]:
common


['ATG',
 'AUS',
 'BRB',
 'CAN',
 'DMA',
 'FJI',
 'GRD',
 'GUY',
 'IND',
 'IRL',
 'JAM',
 'KIR',
 'MLT',
 'MHL',
 'NRU',
 'NZL',
 'PAK',
 'PNG',
 'PHL',
 'KNA',
 'LCA',
 'VCT',
 'WSM',
 'SGP',
 'SLB',
 'TON',
 'TTO',
 'TUV',
 'VUT']

In [69]:
b=[]
for i in range(len(common)):
    if common[i] in a:
        b.append(common[i])
        

In [70]:
b

[]

In [71]:
c=list(a)
len(c)

248

In [72]:
df.head()

Country A2(ISO)   A3 NUM(UN) Dial(cod) eng_spk
0     Afghanistan      AF  AFG       4        93   False
1         Albania      AL  ALB       8       355   False
2         Algeria      DZ  DZA      12       213   False
3  American Samoa      AS  ASM      16     1-684   False
4         Andorra      AD  AND      20       376   False

In [73]:
df[df.NUM(UN) >= 12]

AttributeError: 'DataFrame' object has no attribute 'NUM'